winsup/cygwin/include/cygwin/exit_process.h

```diff
@@ -0,0 +1,348 @@
+ * option to print the address to stdout).
+ *
+ * This function expects the process handle to have the access rights for
+ * CreateRemoteThread(): PROCESS_CREATE_THREAD, PROCESS_QUERY_INFORMATION,
+ * PROCESS_VM_OPERATION, PROCESS_VM_WRITE, and PROCESS_VM_READ.
+ *
+ * The idea for the injected remote thread comes from the Dr Dobb's article "A
+ * Safer Alternative to TerminateProcess()" by Andrew Tucker (July 1, 1999),
+ * http://www.drdobbs.com/a-safer-alternative-to-terminateprocess/184416547.
+ *
+ * The idea to use kernel32!CtrlRoutine for the other signals comes from
+ * SendSignal (https://github.com/AutoSQA/SendSignal/ and
+ * http://stanislavs.org/stopping-command-line-applications-programatically-with-ctrl-c-events-from-net/).
+ */
+
+#include <wchar.h>
+#include <math.h>
+
+#ifndef __INSIDE_CYGWIN__
+/* To help debugging via kill.exe */
+#define small_printf(...) fprintf(stderr, __VA_ARGS__)
+#endif
+
+static LPTHREAD_START_ROUTINE
+kill_via_console_helper(int bitness, wchar_t *function_name, int exit_code, DWORD pid)
+{
+  const char *name;
+  if (bitness == 32)
+    name = "/usr/libexec/getprocaddr32.exe";
+  else if (bitness == 64)
+    name = "/usr/libexec/getprocaddr64.exe";
+  else
+    return NULL; /* what?!? */
+  wchar_t wbuf[PATH_MAX];
+  
+  if (cygwin_conv_path (CCP_POSIX_TO_WIN_W, name, wbuf, PATH_MAX) ||
+      GetFileAttributesW (wbuf) == INVALID_FILE_ATTRIBUTES)
+    return NULL;
+
+  HANDLE h[2];
+  SECURITY_ATTRIBUTES attrs;
+  attrs.nLength = sizeof(attrs);
+  attrs.bInheritHandle = TRUE;
+  attrs.lpSecurityDescriptor = NULL;
+  if (!CreatePipe(&h[0], &h[1], &attrs, 0))
+    return NULL;
+
+  STARTUPINFOW si = {};
+  PROCESS_INFORMATION pi;
+  size_t len = wcslen (wbuf) + wcslen(function_name) + 3 /* exit code */ + 1 /* space */ + 10 /* process ID, i.e. DWORD */ + 1 /* NUL */;
+  WCHAR cmd[len + 1];
+  WCHAR title[] = L"cygwin-console-helper";
+
+  swprintf (cmd, len + 1, L"%S %S %d %ld", wbuf, function_name, exit_code, pid);
+
+  si.cb = sizeof (si);
+  si.dwFlags = STARTF_USESHOWWINDOW | STARTF_USESTDHANDLES;
+  si.wShowWindow = SW_HIDE;
+  si.lpTitle = title;
+  si.hStdInput = si.hStdError = INVALID_HANDLE_VALUE;
+  si.hStdOutput = h[1];
+
+  /* Create a new hidden process.  Use the two event handles as
+     argv[1] and argv[2]. */
+  BOOL x = CreateProcessW (NULL, cmd, NULL, NULL, TRUE,
+			   CREATE_NO_WINDOW | CREATE_NEW_PROCESS_GROUP,
+			   NULL, NULL, &si, &pi);
+  CloseHandle(h[1]);
+  if (!x)
+    {
+      CloseHandle(h[0]);
+      return NULL;
+    }
+
+  CloseHandle (pi.hThread);
+  CloseHandle (pi.hProcess);
+
+  char buffer[64];
+  DWORD offset = 0, count = 0;
+  while (offset < sizeof(buffer) - 1)
+    {
+      if (!ReadFile(h[0], buffer + offset, 1, &count, NULL))
+	{
+	  offset = 0;
+	  break;
+	}
+      if (!count || buffer[offset] == '\n')
+	break;
+      offset += count;
+    }
+  CloseHandle(h[0]);
+
+  buffer[offset] = '\0';
+
+  return (LPTHREAD_START_ROUTINE) strtoull(buffer, NULL, 16);
```

Comment:
Since we're no longer interested in the routine's address, we can save ourselves a lot of trouble, returning the exit code of the spawned process instead. And removing the pump that reads the spawned process' `stdout`. And not even creating that pipe.